packages/sources/ncfx/src/adapter.ts

```diff
@@ -0,0 +1,85 @@
+import { Requester, Validator } from '@chainlink/ea-bootstrap'
+import {
+  Config,
+  ExecuteWithConfig,
+  ExecuteFactory,
+  MakeWSHandler,
+  AdapterRequest,
+} from '@chainlink/types'
+import { makeConfig } from './config'
+
+const inputParams = {
+  endpoint: false,
+}
+
+export const execute: ExecuteWithConfig<Config> = async (request, config) => {
+  const validator = new Validator(request, inputParams)
+  if (validator.error) throw validator.error
+
+  Requester.logConfig(config)
+
+  const jobRunID = validator.validated.id
+
+  return Requester.success(jobRunID, {})
+}
+
+export const makeExecute: ExecuteFactory<Config> = (config) => {
+  return async (request) => execute(request, config || makeConfig())
+}
+
+const customParams = {
+  base: ['base', 'from', 'coin'],
+  quote: ['quote', 'to', 'market'],
+}
+
+export const makeWSHandler = (config?: Config): MakeWSHandler => {
+  const getPair = (input: AdapterRequest) => {
+    const validator = new Validator(input, customParams, {}, false)
+    if (validator.error) return
+    const base = validator.validated.data.base.toUpperCase()
+    const quote = validator.validated.data.quote.toUpperCase()
+    return `${base}/${quote}`
+  }
+  const getSubscription = (request: 'subscribe' | 'unsubscribe', pair?: string) => {
+    if (!pair) return
+    return { request, ccy: pair }
+  }
+  return () => {
+    const defaultConfig = config || makeConfig()
+    return {
+      connection: {
+        url: `${defaultConfig.api.baseWebsocketURL}/cryptodata`,
+      },
+      subscribe: (input) => getSubscription('subscribe', getPair(input)),
+      unsubscribe: (input) => getSubscription('unsubscribe', getPair(input)),
+      subsFromMessage: (message, subscriptionMsg) => {
+        if(Array.isArray(message) && message.length > 0) {
+          const pairMessage = message.find(({ currencyPair }) => currencyPair === subscriptionMsg.ccy)
+          return getSubscription('subscribe', `${pairMessage.currencyPair}`)
+        }
+        return getSubscription('subscribe', `${message}`)
+      },
+      isError: (message: any) => Number(message.TYPE) > 400 && Number(message.TYPE) < 900,
+      filter: () => {
+        return true
+      },
+      toResponse: (message: any, input: any) => {
+        console.log(message)
+        if(Array.isArray(message) && message.length > 0) {
+          const { data: { base, quote } } = input 
+          const pair = `${base}/${quote}`
+          const pairMessage = message.find(({ currencyPair }) => currencyPair === pair)
```

Comment:
Not sure how we should handle when `pairMessage` is null.

Reply:
It will throw on the next line as it tries to walk to result path.
I'd probably just add a log here like:
```
if (!pairMessage) logger.warn(`${pair} not found in message`)
```